### PR TITLE
expression: implement vectorized evaluation for `builtinStrToDateDurationSig`

### PR DESCRIPTION
**Pull request by pingyu**
PCP #12176 

### What problem does this PR solve? 
Implement vectorized evaluation for `builtinStrToDateDurationSig` at #12176 

### What is changed and how it works?
It gets faster about 28.7%.

### Check List 

Tests 

 - Unit test
```
> GO111MODULE=on go test -check.f TestVectorizedBuiltinTimeFunc
PASS: builtin_time_vec_test.go:239: testEvaluatorSuite.TestVectorizedBuiltinTimeFunc	0.157s
PASS: builtin_time_vec_generated_test.go:192: testEvaluatorSuite.TestVectorizedBuiltinTimeFuncGenerated	0.110s
OK: 2 passed
PASS
ok  	github.com/pingcap/tidb/expression	0.292s
```

 - Benchmark
```
> go test -v -benchmem -bench=BenchmarkVectorizedBuiltinTimeFunc -run=BenchmarkVectorizedBuiltinTimeFunc -args builtinStrToDateDurationSig
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinTimeFuncGenerated-12    	1000000000	         0.0165 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinStrToDateDurationSig-VecBuiltinFunc-12         	    1444	    867137 ns/op	  284975 B/op	    3073 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinStrToDateDurationSig-NonVecBuiltinFunc-12      	    1122	    901480 ns/op	  285244 B/op	    3073 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	2.644s
```

### CODE REVIEW

File: expression/builtin_time_vec.go [hunk 1]
```diff
@@ -971,11 +971,61 @@ func (b *builtinExtractDurationSig) vecEvalInt(input *chunk.Chunk, result *chunk
...
+	defer b.bufAllocator.put(bufFormats)
+	if err := b.args[1].VecEvalString(b.ctx, input, bufFormats); err != nil {
+		return err
+	}
+
+	result.ResizeGoDuration(n, false)
+	d64s := result.GoDurations()
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for i := 0; i < n; i++ {
+		if bufStrings.IsNull(i) || bufFormats.IsNull(i) {
```
Comment (qw4990) [inline expression/builtin_time_vec.go]: We can use `result.MergeNulls(bufStrings, bufFormats)`.

Comment (pingyu) [reply expression/builtin_time_vec.go]: Comment addressed. PTAL, Thanks. @qw4990

File: expression/builtin_time_vec.go [hunk 2]
```diff
@@ -1069,11 +1069,61 @@ func (b *builtinExtractDurationSig) vecEvalInt(input *chunk.Chunk, result *chunk
...
+		var t types.Time
+		succ := t.StrToDate(sc, bufStrings.GetString(i), bufFormats.GetString(i))
+		if !succ {
+			if err := handleInvalidTimeError(b.ctx, types.ErrIncorrectDatetimeValue.GenWithStackByArgs(t.String())); err != nil {
+				return err
+			}
+			result.SetNull(i, true)
+			continue
+		}
+		if b.ctx.GetSessionVars().SQLMode.HasNoZeroDateMode() && (t.Time.Year() == 0 || t.Time.Month() == 0 || t.Time.Day() == 0) {
```
Comment (XuHuaiyu) [inline expression/builtin_time_vec.go]: `b.ctx.GetSessionVars().SQLMode.HasNoZeroDateMode()` can be fetched outside of this loop

Comment (pingyu) [reply expression/builtin_time_vec.go]: Comment addressed. PTAL, Thanks. @XuHuaiyu

**Pull request merged**